ethcore/light/src/on_demand/mod.rs

```diff
@@ -413,183 +467,68 @@ impl Handler for OnDemand {
-						Ok(()) => {
-							let _ = sender.send(response.code.clone());
-							return
-						}
-						Err(e) => warn!(target: "on_demand", "Error handling response for code request: {:?}", e),
-					}
-				}
-			}
-			Pending::TxProof(req, sender) => {
-				if let NetworkResponse::Execution(ref response) = *response {
-					match req.check_response(&response.items) {
-						ProvedExecution::Complete(executed) => {
-							let _ = sender.send(Ok(executed));
-							return
-						}
-						ProvedExecution::Failed(err) => {
-							let _ = sender.send(Err(err));
-							return
-						}
-						ProvedExecution::BadProof => warn!(target: "on_demand", "Error handling response for transaction proof request"),
-					}
+				Err(e) => {
+					let peer = ctx.peer();
+					debug!(target: "on_demand", "Peer {} gave bad response: {:?}", peer, e);
+					ctx.disable_peer(peer);
+
+					break;
 				}
 			}
 		}
 
-		ctx.disable_peer(peer);
-	}
+		if pending.requests.is_complete() {
+			let _ = pending.sender.send(pending.responses);
 
-	fn tick(&self, ctx: &BasicContext) {
-		self.dispatch_orphaned(ctx)
-	}
-}
-
-#[cfg(test)]
-mod tests {
-	use super::*;
-
-	use std::sync::Arc;
-
-	use cache::Cache;
-	use net::{Announcement, BasicContext, ReqId, Error as LesError};
-	use request::Requests;
-
-	use network::{PeerId, NodeId};
-	use time::Duration;
-	use util::{H256, Mutex};
-
-	struct FakeContext;
+			return;
+		}
 
-	impl BasicContext for FakeContext {
-		fn persistent_peer_id(&self, _: PeerId) -> Option<NodeId> { None }
-		fn request_from(&self, _: PeerId, _: Requests) -> Result<ReqId, LesError> {
-			unimplemented!()
+		// update network requests (unless we're done, in which case fulfill the future.)
+		let mut builder = basic_request::RequestBuilder::default();
+		let num_answered = pending.requests.num_answered();
+		let mut mapping = move |idx| idx - num_answered;
+
+		for request in pending.requests.requests().iter().skip(num_answered) {
+			let mut net_req = request.clone().into_net_request();
+
+			// all back-references with request index less than `num_answered` have
+			// been filled by now. all remaining requests point to nothing earlier
+			// than the next unanswered request.
+			net_req.adjust_refs(&mut mapping);
+			builder.push(net_req)
+				.expect("all back-references to answered requests have been filled; qed");
 		}
-		fn make_announcement(&self, _: Announcement) { }
-		fn disconnect_peer(&self, _: PeerId) { }
-		fn disable_peer(&self, _: PeerId) { }
-	}
 
-	#[test]
-	fn detects_hangup() {
-		let cache = Arc::new(Mutex::new(Cache::new(Default::default(), Duration::hours(6))));
-		let on_demand = OnDemand::new(cache, 0.into());
-		let result = on_demand.header_by_hash(&FakeContext, request::HeaderByHash(H256::default()));
+		// update pending fields and re-queue.
+		let capabilities = guess_capabilities(&pending.requests.requests()[num_answered..]);
+		pending.net_requests = builder.build();
+		pending.required_capabilities = capabilities;
```

Comment:
Shoudn't you update `pending.requests` here as well (i.e. remove the answered ones?)